File: dataset/test/counts_test.cpp
// SPDX-License-Identifier: GPL-3.0-or-later
// Copyright (c) 2020 Scipp contributors (https://github.com/scipp)
#include "test_macros.h"
#include <gtest/gtest.h>

#include "scipp/dataset/counts.h"
#include "scipp/dataset/dataset.h"

using namespace scipp;
using namespace scipp::dataset;

TEST(CountsTest, toDensity_fromDensity) {
  Dataset d;
  d.setCoord(Dim::Tof, makeVariable<double>(Dims{Dim::Tof}, Shape{4}, units::us,
                                            Values{1, 2, 4, 8}));
  d.setData("", makeVariable<double>(Dims{Dim::Tof}, Shape{3}, units::counts,
                                     Values{12, 12, 12}));

  d = counts::toDensity(std::move(d), Dim::Tof);
  auto result = d[""];
  EXPECT_EQ(result.unit(), units::counts / units::us);
  EXPECT_TRUE(equals(result.values<double>(), {12, 6, 3}));

  d = counts::fromDensity(std::move(d), Dim::Tof);
  result = d[""];
  EXPECT_EQ(result.unit(), units::counts);
  EXPECT_TRUE(equals(result.values<double>(), {12, 12, 12}));
}

File: tests/counts_test.py
# SPDX-License-Identifier: BSD-3-Clause
# Copyright (c) 2023 Scipp contributors (https://github.com/scipp)
import scipp as sc


def test_round_trip_dataset() -> None:
    data = sc.Variable(dims=['tof'], values=[10.0, 10.0], unit=sc.units.counts)
    tof = sc.Variable(dims=['tof'], values=[1.0, 2.0, 3.0], unit=sc.units.us)

    d = sc.Dataset(data={'data': data}, coords={'tof': tof})

    dd = sc.density_to_counts(sc.counts_to_density(d, 'tof'), 'tof')
    assert sc.identical(dd, d)


def test_round_trip_dataarray() -> None:
    data = sc.Variable(dims=['tof'], values=[10.0, 10.0], unit=sc.units.counts)
    tof = sc.Variable(dims=['tof'], values=[1.0, 2.0, 3.0], unit=sc.units.us)

    d = sc.DataArray(data, coords={'tof': tof})

    dd = sc.density_to_counts(sc.counts_to_density(d, 'tof'), 'tof')
    assert sc.identical(dd, d)
